packages/prover/src/proof_provider/payload_store.ts

```diff
@@ -0,0 +1,154 @@
+  private payloads = new Map<BlockELRoot, allForks.ExecutionPayload>();
+
+  private latestBlockRoot: BlockELRoot | null = null;
+
+  constructor(private opts: {api: Api}) {}
+
+  get finalized(): allForks.ExecutionPayload | undefined {
+    const finalizedMaxRoot = this.finalizedRoots.get(this.finalizedRoots.max);
+    if (finalizedMaxRoot) {
+      return this.payloads.get(finalizedMaxRoot);
+    }
+
+    return undefined;
+  }
+
+  get latest(): allForks.ExecutionPayload | undefined {
+    if (this.latestBlockRoot) {
+      return this.payloads.get(this.latestBlockRoot);
+    }
+
+    return undefined;
+  }
+
+  async get(blockId: number | string): Promise<allForks.ExecutionPayload | undefined> {
+    // Given block id is a block hash in hex (32 bytes root takes 64 hex chars + 2 for 0x prefix)
+    if (typeof blockId === "string" && blockId.startsWith("0x") && blockId.length === 64 + 2) {
+      return this.payloads.get(blockId);
+    }
+
+    // Given block id is a block number in hex
+    if (typeof blockId === "string" && blockId.startsWith("0x")) {
+      return this.getOrFetchFinalizedPayload(hexToNumber(blockId));
+    }
+
+    // Given block id is a block number in decimal string
+    if (typeof blockId === "string" && !blockId.startsWith("0x")) {
+      return this.getOrFetchFinalizedPayload(parseInt(blockId, 10));
+    }
+
+    // Given block id is a block number in decimal
+    if (typeof blockId === "number") {
+      return this.getOrFetchFinalizedPayload(blockId);
+    }
+
+    return undefined;
+  }
+
+  async getOrFetchFinalizedPayload(blockNumber: number): Promise<allForks.ExecutionPayload | undefined> {
+    if (blockNumber > this.finalizedRoots.max) {
+      throw new Error(
+        `Block number ${blockNumber} is higher than the latest finalized block number. We recommend to use block hash for unfinalized blocks.`
+      );
+    }
+
+    let blockELRoot = this.finalizedRoots.get(blockNumber);
+    if (!blockELRoot) {
+      const payloads = await getExecutionPayloadForBlockNumber(this.opts.api, this.finalizedRoots.min, blockNumber);
+      for (const payload of Object.values(payloads)) {
+        this.set(payload, true);
+      }
+    }
+
+    blockELRoot = this.finalizedRoots.get(blockNumber);
+    if (blockELRoot) {
+      return this.payloads.get(blockELRoot);
+    }
+
+    return undefined;
+  }
+
+  set(payload: allForks.ExecutionPayload, finalized: boolean): void {
+    const blockRoot = bufferToHex(payload.blockHash);
+    this.payloads.set(blockRoot, payload);
+    this.latestBlockRoot = blockRoot;
+
+    if (finalized) {
+      this.finalizedRoots.set(payload.blockNumber, blockRoot);
+    }
+  }
+
+  async processLCHeader(header: capella.LightClientHeader, finalized = false): Promise<void> {
+    const blockSlot = header.beacon.slot;
+    const blockNumber = header.execution.blockNumber;
+    const blockELRoot = bufferToHex(header.execution.blockHash);
+    const blockCLRoot = bufferToHex(header.beacon.stateRoot);
```

Comment:
can we add a jsdoc here that blockCLRoot is tracking a beacon block by stateRoot

Reply:
Can understand JSDoc for a local variable? Had added doc comments to class members.

Reply:
yes pls add for more clarity to the code reader (or comment)

Reply:
I believe the names of the variables `blockCLRoot, `blockELRoot`, `blockNumber` and `blockSlot` are very clear and self explanatory. No need to add comments for these.